source/server/admin/grouped_stats_request.cc

```diff
@@ -0,0 +1,156 @@
+#include "source/server/admin/grouped_stats_request.h"
+
+#include <string>
+#include <vector>
+
+#include "source/server/admin/prometheus_stats_formatter.h"
+#include "source/server/admin/stats_params.h"
+#include "source/server/admin/stats_render.h"
+
+namespace Envoy {
+namespace Server {
+
+GroupedStatsRequest::GroupedStatsRequest(Stats::Store& stats, const StatsParams& params,
+                                         Stats::CustomStatNamespaces& custom_namespaces,
+                                         UrlHandlerFn url_handler_fn)
+    : StatsRequest(stats, params, url_handler_fn), custom_namespaces_(custom_namespaces),
+      global_symbol_table_(stats.constSymbolTable()) {
+
+  // the "type" query param is ignored for prometheus stats, so always start from
+  // counters; also, skip the TextReadouts phase unless that stat type is explicitly
+  // requested via query param
+  if (params_.prometheus_text_readouts_) {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::TextReadouts, "Text Readouts"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  } else {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  }
+  phase_index_ = 0;
+}
+
+template <class StatType> Stats::IterateFn<StatType> GroupedStatsRequest::saveMatchingStat() {
+  return [this](const Stats::RefcountPtr<StatType>& stat) -> bool {
+    // check if unused
+    if (params_.used_only_ && !stat->used()) {
+      return true;
+    }
+
+    // check if filtered
+    if (params_.filter_ != nullptr) {
+      if (!std::regex_search(stat->name(), *params_.filter_)) {
+        return true;
+      }
+    } else if (params_.re2_filter_ != nullptr &&
+               !re2::RE2::PartialMatch(stat->name(), *params_.re2_filter_)) {
+      return true;
+    }
+
+    // capture stat by either adding to a pre-existing variant or by creating a new variant
+    std::string tag_extracted_name = global_symbol_table_.toString(stat->tagExtractedStatName());
+    auto [iterator, inserted] = stat_map_.try_emplace(
+        tag_extracted_name, std::vector<Stats::RefcountPtr<StatType>>({stat}));
+    if (!inserted) {
```

Comment:
the way you wrote this I think we might get a superfluous allocation when !inserted.

WDYT of inserting an empty vector and appending the stat unconditionally?

It might be slightly faster, both because of 1 less call to the memory allocator and one less branch.

Reply:
Hm I am not sure that will work though. Inserting an empty vector would "reset" any other stats with the same tag extracted name, but different tags

Reply:
From https://en.cppreference.com/w/cpp/container/map/insert

> insert(value_type): Returns a pair consisting of an iterator to the inserted element (or to the element that prevented the insertion) and a bool value set to true if and only if the insertion took place.

So it won't overwrite.

Reply:
aha, got it! will update the code